### PR TITLE
Implement custom section reading/writing

### PR DESCRIPTION
**Pull request by dzfrias**
Draft PR for #1376.

Not fully finished yet, still needs a few things:

- wat2wasm support
- Handling of special custom sections that `BinaryReader` has callbacks for:
  - name
  - dylink0
  - dylink
  - reloc
  - target_features
  - linking
  - code_metadata
- Tests
- Review for consistency with rest of codebase style

I'm not super familiar with all the quirks/idioms of C++, and so I generally followed along with the rest of the codebase. That being said, I could've overlooked some important nuances in this initial draft.

Continuing my conversation with @keithw, I'm still not sure about the best way to handle the custom sections `BinaryReader` has callbacks for. Implementation wise, the backtracking method would be much easier. I go more in-depth on the topic in the second section of [this comment](https://github.com/WebAssembly/wabt/issues/1376#issuecomment-1686707242).

### CODE REVIEW

**Comment by dzfrias** [comment]
I used the backtracking method for reading the special custom sections handled by `BinaryReader` already. Let me know if there are any problems with this implementation.

**Comment by dzfrias** [comment]
I had to update [one of the test inputs](https://github.com/dzfrias/wabt/blob/d965ceb4a8c7119d9081d8498408d5250c1bbafa/test/parse/annotations.txt) to conform with the syntax of the `@custom` annotation.

Let me know if I missed something in this PR!

**Comment by keithw** [comment]
This is looking really nice! Could you please add some roundtrip tests (including for custom sections that use a "place")? I'd also love to see a binary test that shows a custom "name" section for binary names that can't be represented as IDs in the text format, a la https://github.com/WebAssembly/spec/issues/617

It's sort of unfortunate that the annotations proposal itself is never going to supply us with tests for @custom annotations (for reasons discussed at https://github.com/WebAssembly/annotations/issues/15). And while I **wish** this were enough to let us round-trip an LLVM-generated relocatable object from binary to text and back to binary with the linking and relocation sections still valid, BinaryWriter doesn't necessarily write the binary exactly the same (e.g. it may drop the DataCount section). :-(

**Comment by dzfrias** [comment]
> Could you please add some roundtrip tests (including for custom sections that use a "place")?

I'll get working on this! I might have to change a few things, because `WatWriter` writes each custom section at the end of the module, so a completely lossless roundtrip isn't possible with my current implementation.

> I'd also love to see a binary test that shows a custom "name" section for binary names that can't be represented as IDs in the text format, a la https://github.com/WebAssembly/spec/issues/617

Sounds good!

File: src/wast-parser.cc [hunk 1]
```diff
@@ -762,6 +762,13 @@ Result WastParser::ErrorIfLpar(const std::vector<std::string>& expected,
 
 bool WastParser::ParseBindVarOpt(std::string* name) {
   WABT_TRACE(ParseBindVarOpt);
+  if (PeekIsAnnotation("name")) {
```
Comment (dzfrias) [inline src/wast-parser.cc:765]: Is it okay that this is in `ParseBindVarOpt`? I'm not sure if the `@name` annotation is valid in every place that it's used.

**Comment by keithw** [comment]
> On a second look at the other roundtrip tests, is it necessary to have a completely lossless conversion into the `Module`? 

I don't think this is required -- I just meant something like the other roundtrip tests, where it tracks the expected output (so we can tell if it changes in a future commit).

**Comment by dzfrias** [comment]
I wrote a roundtrip test for custom sections. I also roundtripped the `@name` annotation. Let me know if there's anything else I should do!

**Comment by dzfrias** [comment]
No problem; I reverted those commits. I'll open a PR for the `@name` stuff once this one is merged!

File: src/wast-parser.cc [hunk 2]
```diff
@@ -1222,9 +1224,68 @@ Result WastParser::ParseScript(std::unique_ptr<Script>* out_script) {
...
+  Custom custom = Custom(loc, section_name, data);
+  module->customs.push_back(custom);
+
+  return Result::Ok;
+}
+
+bool WastParser::PeekIsCustom() {
+  // If IsLparAnn succeeds, tokens_.front() must have text, as it is an LparAnn
+  // token.
+  return IsLparAnn(PeekPair()) && tokens_.front().text() == "custom";
```
Comment (keithw) [inline src/wast-parser.cc:1279]: Can we throw a `options_->features.annotations_enabled() &&` into this test for safety (and consistency with `PeekMatchRefType`)?

**Review by keithw — APPROVED** [review]
lgtm % comments

@sbc100 do you want to weigh in?

**Comment by keithw** [comment]
Okay, merging -- thank you for contributing this! I'm already enjoying seeing the custom sections from LLVM-produced modules in the wasm2wat output.

**Pull request merged**